generator/altitude_generator.cpp

```diff
@@ -0,0 +1,246 @@
+#include "base/scope_guard.hpp"
+#include "base/stl_helpers.hpp"
+#include "base/string_utils.hpp"
+
+#include "defines.hpp"
+
+#include "std/algorithm.hpp"
+#include "std/type_traits.hpp"
+#include "std/utility.hpp"
+#include "std/vector.hpp"
+
+#include "3party/succinct/elias_fano.hpp"
+#include "3party/succinct/mapper.hpp"
+#include "3party/succinct/rs_bit_vector.hpp"
+
+using namespace feature;
+
+namespace
+{
+using namespace routing;
+
+class SrtmGetter : public IAltitudeGetter
+{
+public:
+  SrtmGetter(string const & srtmDir) : m_srtmManager(srtmDir) {}
+
+  // IAltitudeGetter overrides:
+  feature::TAltitude GetAltitude(m2::PointD const & p) override
+  {
+    return m_srtmManager.GetHeight(MercatorBounds::ToLatLon(p));
+  }
+
+private:
+  generator::SrtmTileManager m_srtmManager;
+};
+
+class Processor
+{
+public:
+  struct FeatureAltitude
+  {
+    FeatureAltitude() : m_featureId(0) {}
+    FeatureAltitude(uint32_t featureId, Altitudes const & altitudes)
+      : m_featureId(featureId), m_altitudes(altitudes)
+    {
+    }
+
+    uint32_t m_featureId;
+    Altitudes m_altitudes;
+  };
+
+  using TFeatureAltitudes = vector<FeatureAltitude>;
+
+  Processor(IAltitudeGetter & altitudeGetter)
+    : m_altitudeGetter(altitudeGetter), m_minAltitude(kInvalidAltitude)
+  {
+  }
+
+  TFeatureAltitudes const & GetFeatureAltitudes() const { return m_featureAltitudes; }
+
+  succinct::bit_vector_builder & GetAltitudeAvailabilityBuilder()
+  {
+    return m_altitudeAvailabilityBuilder;
+  }
+
+  TAltitude GetMinAltitude() const { return m_minAltitude; }
+
+  void operator()(FeatureType const & f, uint32_t const & id)
+  {
+    if (id != m_altitudeAvailabilityBuilder.size())
+    {
+      LOG(LERROR, ("There's a gap in feature id order."));
+      return;
+    }
+
+    bool hasAltitude = false;
+    MY_SCOPE_GUARD(altitudeAvailabilityBuilding, [&] ()
+    {
+      m_altitudeAvailabilityBuilder.push_back(hasAltitude);
+    });
+
+    if (!routing::IsRoad(feature::TypesHolder(f)))
+      return;
+
+    f.ParseGeometry(FeatureType::BEST_GEOMETRY);
+    size_t const pointsCount = f.GetPointsCount();
+    if (pointsCount == 0)
+      return;
+
+    TAltitudes altitudes;
+    bool valid = true;
+    TAltitude minFeatureAltitude = kInvalidAltitude;
+    for (size_t i = 0; i < pointsCount; ++i)
+    {
+      TAltitude const a = m_altitudeGetter.GetAltitude(f.GetPoint(i));
+      if (a == kInvalidAltitude)
+      {
+        valid = false;
+        return;
+      }
+
+      if (minFeatureAltitude == kInvalidAltitude)
+        minFeatureAltitude = a;
+      else
+        minFeatureAltitude = min(minFeatureAltitude, a);
+
+      altitudes.push_back(a);
+    }
+    if (!valid)
+      return;
+
+    hasAltitude = true;
+    m_featureAltitudes.emplace_back(id, Altitudes(move(altitudes)));
+
+    if (m_minAltitude == kInvalidAltitude)
+      m_minAltitude = minFeatureAltitude;
+    else
+      m_minAltitude = min(minFeatureAltitude, m_minAltitude);
```

Comment:
Vladimir Byko-Ianko <notifications@github.com> writes:

> In generator/altitude_generator.cpp:
>
>> +        minFeatureAltitude = a;
>> +      else
>> +        minFeatureAltitude = min(minFeatureAltitude, a);
>> +
>> +      altitudes.push_back(a);
>> +    }
>> +    if (!valid)
>> +      return;
>> +
>> +    hasAltitude = true;
>> +    m_featureAltitudes.emplace_back(id, Altitudes(move(altitudes)));
>> +
>> +    if (m_minAltitude == kInvalidAltitude)
>> +      m_minAltitude = minFeatureAltitude;
>> +    else
>> +      m_minAltitude = min(minFeatureAltitude, m_minAltitude);
>
> Why? If the line is executed m_minAltitude!= kInvalidAltitude and minFeatureAltitude!= kInvalidAltitude. And if the line "minFeatureAltitude = min
> (minFeatureAltitude, a);" is executed minFeatureAltitude!=kInvalidAltitude and a!=kInvalidAltitude.
>
I don't like the idea that kInvalidAltitude may participate in
comparision with others, possibly valid altitudes. This is the case for
min(minFeatureAltitude, m_minAltitude), where m_minAltitude may be
kInvalidAltitude. It's a bad practice to on some "special" parts of
values range. Today we assign uint16_t::min to kInvalidAltitude,
tomorrow we will assign uint16_t::max to kInvalidAltitude, and our
current code'll break. Please, don't do that. This is a lucky
coincidence that kInvalidAltitude is the same as uint16_t::min, but the
code should not depend on this.

> On the other hand kInvalidAltitude is defined as signed int min (not min uint) for the time being.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.*
>

-- 
Best regards, Yuri